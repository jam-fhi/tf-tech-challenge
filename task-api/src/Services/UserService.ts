import PostgresConnection from '../Connection/PostgressConnection';

export interface User {
	first_name: string;
	last_name: string;
	email: string;
	username: string;
	id: number;
}

export default class UserService {
	private dbConn: PostgresConnection;

	constructor(dbConn: PostgresConnection) {
		this.dbConn = dbConn;
	}

	async getUserById(userId: number): Promise<User> {
		const userResult = await this.dbConn.runQuery(
			'select first_name, last_name, email, username, id from public.user where id = $1',
			[userId.toString()]
		);
		if (userResult.rowCount <= 0) {
			throw new Error('No Results');
		}
		return this.buildUser(
			userResult.rows[0].first_name,
			userResult.rows[0].last_name,
			userResult.rows[0].email,
			userResult.rows[0].username,
			userResult.rows[0].id
		);
	}

	async getAllUsers(): Promise<Array<User>> {
		const userResult = await this.dbConn.runQuery(
			'select first_name, last_name, email, username, id from public.user'
		);
		if (userResult.rowCount <= 0) {
			throw new Error('No Results');
		}
		const allUsers: Array<User> = userResult.rows.map((user) => {
			return this.buildUser(
				user.first_name,
				user.last_name,
				user.email,
				user.username,
				user.id
			);
		});
		return allUsers;
	}

	private buildUser(
		first_name: string,
		last_name: string,
		email: string,
		username: string,
		id: number
	): User {
		return {
			first_name,
			last_name,
			email,
			username,
			id,
		};
	}
}
