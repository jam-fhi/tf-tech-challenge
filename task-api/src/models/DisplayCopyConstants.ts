export const SERVER_COPY = {
	LIVE_ON: 'We are live on port',
};
