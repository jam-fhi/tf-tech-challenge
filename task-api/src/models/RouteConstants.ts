export const ALL_USERS: string = 'all_users';
export const BASE: string = 'api';
export const USER: string = 'user';
