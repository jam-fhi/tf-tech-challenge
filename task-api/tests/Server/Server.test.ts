import PostgresConnection from '../../src/Connection/PostgressConnection';
import UserService from '../../src/Services/UserService';
import { setUpUser, tearDownUser, clearUsers } from '../Fixtures/UserFixture';
import { setUpTask, tearDownTask, clearTasks } from '../Fixtures/TaskFixture';
import TaskServer from '../../src/Server/Server';
import superagent from 'superagent';
import { BASE, ALL_USERS, USER } from '../../src/models/RouteConstants';

describe('The server', () => {
	const apiHost = 'http://localhost';
	const host = 'localhost';
	const user = 'pguser';
	const db = 'trustflight';
	const password = 'pgpass';
	const port = 3002;
	const validUserId = 1;
	const invalidUserId = 10;
	const validUsername = 'bsmith';
	const connectFailError = 'connect ECONNREFUSED 127.0.0.1:3002';
	const internalServerError = 'Internal Server Error';

	let dbConn: PostgresConnection;
	let userService: UserService;
	let taskServer: TaskServer;

	beforeEach(async () => {
		process.env.DB_USER = user;
		process.env.DB_HOST = host;
		process.env.DB_PASSWORD = password;
		process.env.DB_DATABASE = db;
		// Due to db seeding, always teardown first.
		await tearDownUser();
		await setUpUser();
		await tearDownTask();
		await setUpTask();
		dbConn = new PostgresConnection();
		userService = new UserService(dbConn);
		taskServer = new TaskServer(userService);
		taskServer.startServer(port);
	});

	afterEach(async () => {
		await taskServer.stopServer();
	});

	it('Will get all users in the system', async () => {
		const result = await superagent.get(
			`${apiHost}:${port}/${BASE}/${ALL_USERS}`
		);
		expect(result.body).toMatchSnapshot();
	});

	it('Will throw an error when there are no users', async () => {
		await clearUsers();
		try {
			await superagent.get(`${apiHost}:${port}/${BASE}/${ALL_USERS}`);
		} catch (e) {
			expect(e.message).toBe(internalServerError);
		}
	});

	it('Will not do anything if the server is already stopped', async () => {
		// Stop called twice so that the if statement
		// within the stop method can be tested
		await taskServer.stopServer();
		await taskServer.stopServer();
		try {
			await superagent.get(`${apiHost}:${port}/${BASE}/${ALL_USERS}`);
		} catch (e) {
			expect(e.message).toBe(connectFailError);
		}
	});

	it('Will get users with a valid id', async () => {
		const result = await superagent.get(
			`${apiHost}:${port}/${BASE}/${USER}?userid=${validUserId}`
		);
		expect(result.body.username).toBe(validUsername);
	});

	it('Will fail to get users with a valid id', async () => {
		clearUsers();
		try {
			await superagent.get(
				`${apiHost}:${port}/${BASE}/${USER}?userid=${validUserId}`
			);
		} catch (e) {
			expect(e.message).toBe(internalServerError);
		}
	});

	it('Will fail to get users an invalid user id', async () => {
		clearUsers();
		try {
			await superagent.get(
				`${apiHost}:${port}/${BASE}/${USER}?userid=${invalidUserId}`
			);
		} catch (e) {
			expect(e.message).toBe(internalServerError);
		}
	});

	it('Will fail to get users with no userid query', async () => {
		clearUsers();
		try {
			await superagent.get(`${apiHost}:${port}/${BASE}/${USER}`);
		} catch (e) {
			expect(e.message).toBe(internalServerError);
		}
	});
});
