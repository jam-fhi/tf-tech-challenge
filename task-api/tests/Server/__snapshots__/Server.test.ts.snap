// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The server Will get all tasks in the system 1`] = `
Array [
  Object {
    "id": 1,
    "taskName": "Make a cup of tea",
    "userid": 1,
  },
  Object {
    "id": 2,
    "taskName": "Check the left wing",
    "userid": 2,
  },
]
`;

exports[`The server Will get all users in the system 1`] = `
Array [
  Object {
    "email": "Bob@smith.comm",
    "first_name": "Bob",
    "id": 1,
    "last_name": "Smith",
    "username": "bsmith",
  },
  Object {
    "email": "mary@jones.comm",
    "first_name": "Mary",
    "id": 2,
    "last_name": "Jones",
    "username": "mjones",
  },
]
`;
